antsibull/utils/http.py

```diff
@@ -0,0 +1,49 @@
+# Author: Felix Fontein <felix@fontein.de>
+# License: GPLv3+
+# Copyright: Ansible Project, 2020
+"""General functions for working with aiohttp."""
+
+import asyncio
+import typing as t
+
+import aiohttp
+
+
+def _format_call(command: str, args: t.Tuple[t.Any, ...], kwargs: t.Mapping[str, t.Any]) -> str:
+    arguments = [repr(a) for a in args] + ['{0}={1}'.format(k, repr(v)) for k, v in kwargs.items()]
+    return 'aio_session.{command}({args})'.format(
+        command=command,
+        args=', '.join(arguments),
+    )
+
+
+async def retry_get(aio_session: 'aiohttp.client.ClientSession',
+                    *args,
+                    acceptable_error_codes: t.Optional[t.Iterable[int]] = None,
+                    max_retries: int = 10,
+                    **kwargs):
+    error_codes = []
+    for retry in range(max_retries):
```

Comment:
Okay, so I said earlier that I thought max_retries, or at least the dfault value of max_retries should us the new config file + context stuff that I wrote.  I'm not completely happy with the API of all of that but let me see if I can explain how to do it (and maybe you'll think of some better ways to rewrite some of that after you have to do it once ;-)  The way that will work is:

**Setup**

* Add a new config option, `max_retries` to the ConfigModel https://github.com/ansible-community/antsibull/blob/main/antsibull/config.py#L91
  * At the moment, this schema is only used for validating a config file.  The defaults only matter in that they should agree with the type and they need to exist so that a config file without that entry defined still validates.  This is a wart but I'm not sure how I want to fix it yet.
* Add `max_retries` to the `_FIELDS_IN_LIB_CTX` set: https://github.com/ansible-community/antsibull/blob/main/antsibull/app_context.py#L76  It determines which fields in args and config will get placed into the "schema-ized" section of the context objects vs ones which will be added to the free-for-all sections.  Everything that makes it into the lib_ctx should be schema-ized.
* We're adding this to the lib_ctx so add it in here: https://github.com/ansible-community/antsibull/blob/main/antsibull/app_context.py#L232  The default value here is what matters.
* You can also add it to the sample antsibull.cfg file: https://github.com/ansible-community/antsibull/blob/main/antsibull.cfg  (Someday in the future I should add a command to generate a config file with all the defaults.... That should be something like `perky.dump(filename, ConfigModel().dict())`but I haven't thought about how to plumb it to the command line args of the various scripts.... 
* You'll also have to modify the unit tests for context to know about the new value: https://github.com/ansible-community/antsibull/blob/main/tests/units/test_context.py  This should probably be structured differently... right now adding a new value means adding the new value to most of the unittests in there... Some sort of abstraction, even if it just allowed you to enter the new value a single time and all the tests would then check that the default value had been set on the context would be good.

**Usage**
* In `retry_get()`, you'll want to to do something like this: https://github.com/ansible-community/antsibull/blob/main/antsibull/galaxy.py#L207
* First retrieve the current lib_ctx with something like: `lib_ctx = app_context.lib_ctx.get()`
* Then make use of it in the code with something like:
  ``` python
  if max_retries is None:
      max_retries = lib_ctx.max_retries
    ```

Reply:
Implemented in 5479f577d2f5b87c6ab798b7f73af17f28ac1138.